notebooks/utils.py

```diff
@@ -0,0 +1,127 @@
+            'nepoch': 100,
+            'lr': 0.002,
+            'momentum': 0.9,
+            'swarm_size': 20
+            
+        }
+        
+        self.simulations = []
+        
+        
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        
+        
+    def reset_swarm(self):
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+    def rebuild_predictor(self):
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        self.reset_swarm()
+        
+
+    def make_net(self, hidden_depth, width):
+        assert hidden_depth >= 1
+        yield nn.Linear(1, width)
+        yield nn.ReLU()
+        for i in range(hidden_depth-1):
+            yield nn.Linear(width, width)
+            yield nn.ReLU()
+        yield nn.Linear(width, 1)
+    
+    def make_predictor(self, x, y):
+        xt = x.clone()
+        yt = y.clone()
+        def get_pred(num, hidden = self.params['hidden'], width = self.params['width'], nepoch=self.params['nepoch'], lr=self.params['lr'], momentum=self.params['momentum'], debug=self.debug):
+            net = nn.Sequential(*self.make_net(hidden, width))
+            lossfunc = nn.MSELoss()
+            optimizer = optim.SGD(net.parameters(), lr=lr, momentum=momentum)
+
+            og_loss = lossfunc(net(xt), yt)
+            for epoch in range(nepoch):
+                optimizer.zero_grad()
+                ypred = net(xt)
+                self.bees[num]['pred_list'].append(np.array(ypred.detach()))
+                loss = lossfunc(ypred, yt)
+                self.bees[num]['loss_list'].append(loss.item())
+                if debug: print(epoch, loss)
+                loss.backward()
+                optimizer.step()
+                
+            if debug: print(f"First loss {og_loss} v final {loss}")
+            return ypred.detach(), loss.item()
+        return get_pred
+
+    def run_sim(self, debug = False):
+        
+                  
+        for num in range(len(self.bees)):
+            yval, loss = self.get_pred(num)
+            if np.isnan(loss):
+                raise RuntimeError("Nan loss")
+            if debug:
+                print(num, loss)
+        
+        sim_dict = {
```

Comment:
This is definitely a good thing to store! We can then visualise of identical experiments instead of purely random results each time. I was thinking of doing it via a seed, but this is good too!

Reply:
Oh thanks! (finally I did something commendable 😆 ) Yeah my whole goal was to be able to save this, particularly with the params. GIven that animation is so much slower than actually running experiments, and animation won't be the only way of doing analysis I think it doesn't make much sense to handcuff the generation of results with visualising them in an animation. Animations are kinda a dead-end if you don't have the exact inputs stored somewhere, or reproducable. I should have just worked on your animation module to make that happen.